pkg/util/subscriptions/publisher.go

```diff
@@ -0,0 +1,23 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package subscriptions
+
+import "go.uber.org/fx"
+
+// Publisher represents a component's request for a transmitter of this type.
+//
+// A component's constructor should take an object of this type as an argument
+// (or indirectly via an fx.In struct), and then call its Transmitter method.
+type Publisher[M Message] struct {
```

Comment:
I have tried to reduce the number of types the user has to handle and come the following solution
It is simpler to use (Less types to remember) but the internal implementation is more visible to the user (encapsulation is not so great)

```go
type Publisher[M Message] struct {
	fx.In

	Receivers []chan M `group:"receivers"`
}

func (p Publisher[M]) Notify(message M) {
	for _, rev := range p.Receivers {
		rev <- message
	}
}

type Receiver[M Message] struct {
	fx.Out
	Ch chan M `group:"receivers"`
}

func NewReceiver[M Message]() Receiver[M] {
	return Receiver[M]{
		Ch: make(chan M, 1),
	}
}

func (s Receiver[M]) Chan() <-chan M {
	return s.Ch
}
```
What do you think about this solution?

Reply:
That is, indeed, simpler.  I had been concerned that `fx.In` and `fx.Out` would occupy unused space at runtime, but this does not appear to be the case -- they are both ZSTs.  The only functionality missing in your implementation is filtering `nil`s out of `p.Receivers`.  But probably that could be done in `Notify` with no real impact on performance.

So, I'll adopt this -- thanks!